### DISCUSSION TITLE
[Q] firewalld and nftables and docker

### DISCUSSION THREAD

**Discussion by lonix1** [firewalld/firewalld#1125] [Q&A] [answered]
I've encountered countless articles/tutorials/whatever stating that firewalld and docker do not work well in tandem, because firewalld now uses nftables instead of iptables, but docker only works with iptables. Or something like that. So it's recommended to avoid firewalld, or use it in a compatibility mode where it uses legacy iptables instead.

Then I've seen posts saying that this isn't true.

And on the firewalld [homepage](https://firewalld.org), it says:
> Applications and libraries which support firewalld as a firewall management tool include:
>    docker (iptables backend only)

It's all really quite confusing.

As of today, do firewalld, nftables and docker work well together?

I'd really like to use it because I don't feel like learning nftables. Thanks for your insights and help.

**Comment by erig0** [comment]
This is a complex answer because the situation has changed many times throughout history. But I can answer it in regards to _today_. :)

> As of today, do firewalld, nftables and docker work well together?

Yes.

For one caveat see #869. Docker wants to open port forwarding (`-p`, `--publish`) to all interfaces. Firewalld will honor this. That surprises some users.

**Reply by lonix1** [reply]
Very interesting. Even the docker docs don't mention nftables - I thought docker only works with iptables... I'm very surprised. Good news indeed.

**Reply by erig0** [reply, marked as answer]
docker uses iptables.
firewalld can use either iptables or nftables.

All combinations should work.
docker + firewalld iptables
docker + firewalld nftables

**Reply by lonix1** [reply]
Thanks, and please excuse my ignorance. I've read MANY articles and StackOverflow posts and so was pushed to believe something else. It's good to read the truth from the horse's mouth!

Am I right in my understanding, that if docker uses only iptables, but firewalld works with both, then firewalld acts as a facade for nftables. So docker doesn't choke on the config because as far as it's concerned it doesn't care about the implementation - it integrates with firewalld directly and ignores the rest?

So as long as I keep you above caveat in mind, I can FINALLY! switch to nftables.

**Reply by polarathene** [reply]
Docker has some integration with firewalld (_with it's own `docker` zone since 2020H2, previously advising to add `docker0` to `trusted` zone_), but for the most part it just manages it's own rules with iptables (_which you probably have a package installed that actually rewrites these as rules for nftables under the hood_).

You're right about the confusing information out there. Even in 2023, you'll find some content being produced that cites old configuration with adding `docker0` to the `trusted` zone.

---

Not much changes with firewalld in the mix AFAIK, except:

- As mentioned containers that use the port publishing feature `-p` of docker does "bypass" the firewall (_both for firewalld or UFW_), but this only takes users by surprise that expect the firewall to prevent that without explicitly permitting it through the firewall frontend (firewalld / UFW).
  - You can avoid using this docker feature, and instead manually use `firewall-cmd` to forward a public port from an interface (or zone rather in firewalld?) to the container IP + port
  - Alternatively just publish to an explicit interface `-p 127.0.0.1:80:8080` (internal only), or set the default bind interface in the `/etc/docker/daemon.json` config and only provide an explicit interface (_like your public IP, or all interfaces `0.0.0.0`_) when you want it to be public accessible.
  - **NOTE:** private interfaces on the docker host may still be reachable from other hosts on a LAN, firewalld can partially mitigate that with the zones feature (_see below for more info_).

---

The only surprise I've experienced so far was restarting or stopping the firewalld service wiped the `iptables` rules Docker creates, which affected connectivity until restarting the Docker daemon.
- Docker does have integration for reacting to a _reload event_ emitted from firewalld (`firewalld-cmd --reload`), which will recreate those rules (_but not if firewalld was stopped or restarted AFAIK, that requires restarting docker_).
- If you have added any custom rules (eg: `DOCKER-USER` chain), these won't be preserved AFAIK, but an upcoming release of firewalld will soon leave those alone instead of deleting them.

---

### Hosts on a LAN may be able to access bound ports to a hosts private networks

This issue has [been reported some time ago to the Docker (moby) project](https://github.com/moby/moby/issues/14041#issuecomment-223467379). I am not familiar with networking enough, but some have suggested that it could be used if you are connected to an untrusted network like at a Cafe / Airport wifi with a work laptop? I think moby is working on an alternative networking solution with IPVS which may avoid the issues :man_shrugging: 

Partial mitigation seems to be only applicable to firewalld thanks to it's use of zones to manage interfaces? (_An advantage over UFW_)

A host on the same LAN can adjust their routing to access the docker hosts private networks (_eg: `127.0.0.0/8` . `172.17.0.0./16`_).
- This is possible due to Docker enabling `sysctl net.ipv4.ip_forward=1` which it uses to allow external hosts connect to containers via published ports. Dockers `iptables` forwarding rules then allow traffic to a container via the published port, even when you explicitly bind to private networks only.
- For a container with a published port to the docker hosts IPv4 `localhost` (_with `-p 127.0.0.1:80:8080`_), the attacker can connect to the published container port:

   ```bash
   ip addr add 127.0.0.2/8 dev lo
   ip addr del 127.0.0.1/8 dev lo
   ip route 127.0.0.1 via <target LAN IP here>
   ip route 172.17.0.0/16 via <target LAN IP here>
   
   # Example curl targets: `127.0.0.1:80` or `172.17.0.2:8080`
   
   # NOTE: With docker network (docker0) isolated from the LAN interface
   # in the docker zone, the container IP is not routable thanks to firewalld.
  ```

I have reproduced it with two VMWare guests (Arch Linux) in a NAT network, and with Vultr VPC (_between two instances: Fedora firewalld, Ubuntu UFW_):
- You can then do something like a port scan with `nmap` to find what IPs/ports available to connect to. It will only be the published ports on `127.0.0.1` (_nothing else listening on `127.0.0.1` outside of docker_), or the equivalent container port at the containers IP (_only if published, non-published ports remain unreachable_).
- If `docker0` is moved from the `docker` zone to the zone with the interface for the LAN IP, then the `172.17.0.0/16` subnet becomes reachable with firewalld too (_which makes sense why UFW doesn't mitigate that access_).
- AFAIK this is [routing between two Layer 3 networks via Layer 2 network switch?](https://serverfault.com/questions/696433/differences-between-adding-ip-route-to-system-and-adding-additional-ip-address/696437#696437), hence why it only works for LAN networks, and not servers interface with public IP? (_or that would require more routing at the target host as well?_ :man_shrugging: )

Perhaps it's possible to also prevent the `127.0.0.1` routing, but I have no clue :sweat_smile:

**Reply by erig0** [reply]
> The only surprise I've experienced so far was restarting or stopping the firewalld service wiped the iptables rules Docker creates, which affected connectivity until restarting the Docker daemon.

This is no longer occurs as of #1082. Firewalld will avoid touching iptables if `FirewallBackend=nftables` (the default) and there are no `--direct` rules (the default).

**Reply by lonix1** [reply]
@polarathene Thank you for the detailed analysis! Getting all this to work is really tough, because as you highlighted, most content out there refers to old code.

One follow-up question if I may: Like many people, I run a bunch of containers on my server, and use a reverse proxy to serve them. So (other than port 22) only ports 80 and 443 are open publicly, and the containers interact with them only. I assume the "gotchas" and issues you detailed above don't apply in this simplistic use case?

**Reply by polarathene** [reply]
@lonix1 I [opened a separate discussion for the LAN hosts concern](https://github.com/firewalld/firewalld/discussions/1128) (_which firewalld helps a bit with the special zone for Docker_), and added a `iptables` command that fixes the remaining issue with the LAN hosts routing to `127.0.0.1`.

> I assume the "gotchas" and issues you detailed above don't apply in this simplistic use case?

For ports intended to be public, it's a non-issue.

If you only publish those two ports and have it route to containers within it's own docker network is fine. The only gotcha here is the firewall (firewalld/UFW) doesn't have any say over allowing/denying access to those containers published ports, so long as you are aware of that, nothing to worry about :+1: 

One user was doing similar as you, but they also made the containers connected to the reverse-proxy reachable via published ports on `127.0.0.1` for local dev / testing, which exposed some unexpected risk (eg: laptop at cafe / airport wifi).

As long as `iptables --list | grep FORWARD` has "policy DROP", the other risk of LAN hosts connecting to containers without ports published is a non-issue. It should be `DROP` by default when Docker is running, but in some cases something else may have changed that.
